src/implementations/LinearAlgebra.jl

```diff
@@ -155,6 +155,13 @@ function similar_array_type(
     return LinearAlgebra.Symmetric{S,similar_array_type(MT, S)}
 end
 
+function similar_array_type(
+    ::Type{LinearAlgebra.Diagonal{T,VT}},
+    ::Type{S},
+) where {S,T,VT<:AbstractVector{T}}
+    return LinearAlgebra.Diagonal{S,Vector{S}}
```

Comment:
So I tried this, but it doesn't work for a bunch of types. One example is `1.2 * Diagonal(1:2)`, because `1.2 * (1:2)` is a nasty type. Forcing a plain vector here seemed best.

Reply:
If the default `similar_array_type(::Type{<:AbstractVector}, ::Type{T}} = Vector{T}` would make it work for your range example